crypto/s2n_comp_cipher_aes_sha.c

```diff
@@ -0,0 +1,265 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
+#include "crypto/s2n_cipher.h"
+
+#include "tls/s2n_crypto.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_blob.h"
+
+static int s2n_comp_cipher_aes_sha_initial_hmac(struct s2n_session_key *key, uint8_t *sequence_number, uint8_t content_type,
+                                                uint16_t protocol_version, uint16_t payload_and_eiv_len, int *extra)
+{
+    uint8_t ctrl_buf[EVP_AEAD_TLS1_AAD_LEN];
+    struct s2n_blob ctrl_blob = { .data = ctrl_buf, .size = EVP_AEAD_TLS1_AAD_LEN };
+    struct s2n_stuffer ctrl_stuffer;
+    GUARD(s2n_stuffer_init(&ctrl_stuffer, &ctrl_blob));
+
+    GUARD(s2n_stuffer_write_bytes(&ctrl_stuffer, sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    GUARD(s2n_stuffer_write_uint8(&ctrl_stuffer, content_type));
+    GUARD(s2n_stuffer_write_uint8(&ctrl_stuffer, protocol_version / 10));
+    GUARD(s2n_stuffer_write_uint8(&ctrl_stuffer, protocol_version % 10));
+    GUARD(s2n_stuffer_write_uint16(&ctrl_stuffer, payload_and_eiv_len));
+
+    /* This will unnecessarily mangle the input buffer, which is fine since we're using a temporary buffer
+     * Return value will be length of digest, padding, and padding length byte.
+     * See https://github.com/openssl/openssl/blob/master/crypto/evp/e_aes_cbc_hmac_sha1.c#L814
+     * and https://github.com/openssl/openssl/blob/4f0c475719defd7c051964ef9964cc6e5b3a63bf/ssl/record/ssl3_record.c#L743
+     */
+    int ctrl_ret = EVP_CIPHER_CTX_ctrl(&key->native_format.evp_cipher_ctx, EVP_CTRL_AEAD_TLS1_AAD, EVP_AEAD_TLS1_AAD_LEN, ctrl_buf);
+
+    if (ctrl_ret < 0) {
+        S2N_ERROR(S2N_ERR_INITIAL_HMAC);
+    }
+
+    *extra = ctrl_ret;
+    return 0;
+}
+
+static int s2n_comp_cipher_aes_sha_encrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *in, struct s2n_blob *out)
+{
+    eq_check(out->size, in->size);
+
+    if (EVP_EncryptInit_ex(&key->native_format.evp_cipher_ctx, NULL, NULL, NULL, iv->data) == 0) {
+        S2N_ERROR(S2N_ERR_KEY_INIT);
+    }
+
+    if (EVP_Cipher(&key->native_format.evp_cipher_ctx, out->data, in->data, in->size) == 0) {
+        S2N_ERROR(S2N_ERR_ENCRYPT);
+    }
+
+    return 0;
+}
+
+static int s2n_comp_cipher_aes_sha_decrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *in, struct s2n_blob *out)
+{
+    eq_check(out->size, in->size);
+
+    if (EVP_DecryptInit_ex(&key->native_format.evp_cipher_ctx, NULL, NULL, NULL, iv->data) == 0) {
+        S2N_ERROR(S2N_ERR_KEY_INIT);
+    }
+
+    if (EVP_Cipher(&key->native_format.evp_cipher_ctx, out->data, in->data, in->size) == 0) {
+        S2N_ERROR(S2N_ERR_DECRYPT);
+    }
+
+    return 0;
+}
+
+static int s2n_comp_cipher_aes_sha_get_mac_write_key(struct s2n_session_key *key, uint8_t *mac_key, uint32_t mac_size)
```

Comment:
Yep, I'll change get_{encryption, decryption}_key to set_{encryption,decryption}_key as well.